collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java

```diff
@@ -0,0 +1,149 @@
+                {"regularGroupWithQuestionsAndRegularGroupInside.xml", new String[] {"-1, ", "0, 0, ", "0, 1, 0, ", "0, 2, ", "-1, ", "0, 2, "}},
+                {"fieldListWithQuestionAndRegularGroupInside.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"fieldListWithQuestionsAndRegularGroupsInside.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"threeNestedFieldListGroups.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}}
+        });
+    }
+
+    private String formName;
+    private String[] expectedIndices;
+
+    public FormNavigationTestCase(String formName, String[] expectedIndices) {
+        this.formName = formName;
+        this.expectedIndices = expectedIndices;
+    }
+
+    @Test
+    public void formNavigationTestCase() throws ExecutionException, InterruptedException {
+        testIndices(formName, expectedIndices);
+    }
+
+    private void testIndices(String formName, String[] expectedIndices) throws ExecutionException, InterruptedException {
+        try {
+            prepareFile(formName);
+        } catch (IOException e) {
+            Timber.i(e);
+        }
+        FormLoaderTask formLoaderTask = new FormLoaderTask(formPath(formName), null, null);
+        formLoaderTask.setFormLoaderListener(new FormLoaderListener() {
+            @Override
+            public void loadingComplete(FormLoaderTask task) {
+                try {
+                    // For each form, simulate swiping forward through screens until the end of the
+                    // form and then swiping back once. Verify the expected indices before and after each swipe
+                    for (int i = 0; i < expectedIndices.length - 1; i++) {
+                        FormController formController = task.getFormController();
+                        // check the current index
+                        assertEquals(expectedIndices[i], formController.getFormIndex().toString());
+                        if (i < expectedIndices.length - 2) {
+                            formController.stepToNextScreenEvent();
+                        } else {
+                            formController.stepToPreviousScreenEvent();
+                        }
+                        // check the index again after navigating
+                        assertEquals(expectedIndices[i + 1], formController.getFormIndex().toString());
+                    }
+                } catch (Exception e) {
+                    Timber.i(e);
+                }
+            }
+
+            @Override
+            public void loadingError(String errorMsg) {
+            }
+
+            @Override
+            public void onProgressStep(String stepMessage) {
+
+            }
+        });
+        formLoaderTask.execute(formPath(formName)).get();
+    }
+
+    /*
+    FormLoaderTask loads forms from SD card so we need to put each form there
+     */
+    private void prepareFile(String formName) throws IOException {
```

Comment:
This could be a doc comment ( /** ... */ ). Might the method have a more descriptive, specific name? `copyToSdCard`, perhaps?